File: js/src/typed/nodes/FileNode.ts
// Copyright (c) 2018 Spotify AB.
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

/* Generated */

import { v4 as uuid } from 'uuid';
import { Command, Node, ContentType } from '../../score';
import ArgMapper from '../../schema/ArgMapper';
import TypedNode from './TypedNode';

/**
 * FileNodeConfig.
 */
export interface FileNodeConfig {
  file: string;
  when?: number;
  duration?: number;
  offset?: number;
}

/**
 * FileNode.
 */
class FileNode extends TypedNode {
  /**
   * Unique identifier for the plugin kind this node represents.
   */
  public static PLUGIN_KIND = 'com.nativeformat.plugin.file.file';

  /**
   * Creates a new `string` mapper for the `file` param.
   */
  private static FILE_CONFIG = ArgMapper.newStringArg('file', undefined);

  /**
   * Creates a new `number` mapper for the `when` param.
   */
  private static WHEN_CONFIG = ArgMapper.newTimeArg('when', 0);

  /**
   * Creates a new `number` mapper for the `duration` param.
   */
  private static DURATION_CONFIG = ArgMapper.newTimeArg('duration', 0);

  /**
   * Creates a new `number` mapper for the `offset` param.
   */
  private static OFFSET_CONFIG = ArgMapper.newTimeArg('offset', 0);

  /**
   * Creates a new `FileNode` instance.
   */
  constructor(
    id: string,
    public readonly file: string,
    public readonly when: number,
    public readonly duration: number,
    public readonly offset: number
  ) {
    super(id, FileNode.PLUGIN_KIND);
  }

  /**
   * `FileNode` factory.
   */
  public static create(config: FileNodeConfig, id: string = uuid()): FileNode {
    return new FileNode(
      id,
      FileNode.FILE_CONFIG.getValueOrThrow(config.file),
      FileNode.WHEN_CONFIG.getValueOrThrow(config.when),
      FileNode.DURATION_CONFIG.getValueOrThrow(config.duration),
      FileNode.OFFSET_CONFIG.getValueOrThrow(config.offset)
    );
  }

  /**
   * Creates a new `FileNode` from a score `Node`.
   */
  public static from(node: Node): FileNode {
    if (FileNode.PLUGIN_KIND !== node.kind) {
      throw new Error(`Expected plugin kind="${FileNode.PLUGIN_KIND}"`);
    }
    return new FileNode(
      node.id,
      FileNode.FILE_CONFIG.readConfig(node),
      FileNode.WHEN_CONFIG.readConfig(node),
      FileNode.DURATION_CONFIG.readConfig(node),
      FileNode.OFFSET_CONFIG.readConfig(node)
    );
  }

  /**
   * Configs for this node.
   */
  public getConfig(): FileNodeConfig {
    return {
      file: this.file,
      when: this.when,
      duration: this.duration,
      offset: this.offset
    };
  }

  /**
   * Inputs for this node.
   */
  public getInputs(): Map<string, ContentType> {
    return new Map([]);
  }

  /**
   * Outputs for this node.
   */
  public getOutputs(): Map<string, ContentType> {
    return new Map([['audio', ContentType.AUDIO]]);
  }
}

/**
 * Export.
 */
export default FileNode;

File: swift/Sources/Grapher/Nodes/FileNode.swift
// Copyright (c) 2018 Spotify AB.
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License

/* Generated */

import Foundation

/// A plugin that plays different types of media files.
public final class FileNode: Node {

    private enum ConfigKeys: String, CodingKey {
        case file = "file"
        case when = "when"
        case duration = "duration"
        case offset = "offset"
    }

    public override class var inputs: [String: ContentType] {
        return [:]
    }

    public override class var outputs: [String: ContentType] {
        return ["audio": .audio]
    }

    /// Tells the plugin where to pull the file from, it supports both local files and HTTP based files.<br>Possible string formats:</br><br>- File: `<absolute-path-to-file>`</br><br>- Streamable Audio: `<HTTP(s)-URL>`</br><br>- Spotify Track: `spotify:track:<uri>`</br><br>- MIDI File: `midi:<absolute-path-to-midi-file>:soundfont:<absolute-path-to-soundfont-file>`
    public var file: String

    /// Tells the plugin when to start playing the file.
    public var when: Time

    /// Tells the plugin how long to play the file for.
    public var duration: Time

    /// Tells the plugin where within the track playback should begin.
    public var offset: Time

    /**
        Designated initializer

        - Parameters:
            - id: Unique identifier of this node.
            - file: Tells the plugin where to pull the file from, it supports both local files and HTTP based files.<br>Possible string formats:</br><br>- File: `<absolute-path-to-file>`</br><br>- Streamable Audio: `<HTTP(s)-URL>`</br><br>- Spotify Track: `spotify:track:<uri>`</br><br>- MIDI File: `midi:<absolute-path-to-midi-file>:soundfont:<absolute-path-to-soundfont-file>`
            - when: Tells the plugin when to start playing the file.
            - duration: Tells the plugin how long to play the file for.
            - offset: Tells the plugin where within the track playback should begin.
     */
    public init(id: String = IDGenerator.generateUniqueID(), file: String, when: Time = Time(nanos: 0), duration: Time = Time(nanos: 0), offset: Time = Time(nanos: 0)) {
        self.file = file
        self.when = when
        self.duration = duration
        self.offset = offset

        super.init(id: id, kind: .fileNode)
    }

    // MARK: Codable

    public required init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let config = try container.nestedContainer(keyedBy: ConfigKeys.self, forKey: .config)
        file = try config.decode(String.self, forKey: .file)
        when = try config.decodeIfPresent(Time.self, forKey: .when) ?? Time(nanos: 0)
        duration = try config.decodeIfPresent(Time.self, forKey: .duration) ?? Time(nanos: 0)
        offset = try config.decodeIfPresent(Time.self, forKey: .offset) ?? Time(nanos: 0)

        try super.init(from: decoder)
    }

    public override func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        var config = container.nestedContainer(keyedBy: ConfigKeys.self, forKey: .config)
        try config.encode(file, forKey: .file)
        try config.encode(when, forKey: .when)
        try config.encode(duration, forKey: .duration)
        try config.encode(offset, forKey: .offset)

        try super.encode(to: encoder)
    }
}
